tests/instrument/expr_match.ml.reference

```diff
@@ -99,3 +99,174 @@ let f x =
         (match x with
          | Foo  -> (___bisect_mark___expr_match 36; print_endline "foobar")
          | Bar  -> (___bisect_mark___expr_match 37; print_endline "barbar"))))
+let f x =
+  ___bisect_mark___expr_match 43;
+  (match x with
+   | Foo |Bar  as ___bisect_matched_value___ ->
+       ((((match ___bisect_matched_value___ with
+           | Foo  -> (___bisect_mark___expr_match 41; ())
+           | Bar  -> (___bisect_mark___expr_match 42; ())
+           | _ -> ()))[@ocaml.warning "-8-11"]);
+        print_endline "foo"))
```

Comment:
I think you meant that we'll cause the side effect twice if _both_ patterns match; but I think that is exactly what we want. My understanding of `or` patterns in `OCaml` is that there should not be a difference between your two previous examples, and that the first form is just a shorter (more convenient) way for of writing the second.

With regard to promoting expressions, again, I'm not so certain that this would be that difficult. As I understand it, both forms of the expression would receive the same "closure" (in quotes because I'm not certain I'm being precise; I mean the set of all bound variables), so no extra search is necessary.

I completely agree that the way you implemented `translate` this should be easy. I unfortunately, don't have the time to dive in and try implementing that approach. I've tested this enough to feel confident that this works so we can merge, but I'd like 3 things:

1. can we add a command line switch to disable this "or" expansion ? I think that this is a powerful tool and should be on by default, but a switch might help to easily diagnose weird corner cases that we might miss.
2. can you also add the above explanation to the docs (README.md), it is pretty good.
3. I want to file an issue, for myself, to write the "or" expansion with expression promotion. It will help me to understand the problem better, and we can not merge if there is no significant improvement.